src/porepy/numerics/ad/operators.py

```diff
@@ -1234,15 +1132,228 @@ def _parse_other(self, other):
+            raise TypeError(
+                "Operator functions must be called before applying any operation."
+            )
+        elif isinstance(other, Operator):
+            # Put Operator at end, because Seconary and Abstract are also operators
+            return [self, other]
         else:
             raise ValueError(f"Cannot parse {other} as an AD operator")
 
 
+class TimeDependentOperator(Operator):
+    """Intermediate parent class for operator classes, which can have a time-dependent
+    representation.
+
+    Implements the notion of time step indices, as well as a method to create a
+    representation of an operator instance at a previous time.
+
+    Operators created via constructor always start at the current time.
+
+    """
+
+    def __init__(
+        self,
+        name: str | None = None,
+        domains: Optional[pp.GridLikeSequence] = None,
+        operation: Optional[Operator.Operations] = None,
+        children: Optional[Sequence[Operator]] = None,
+    ) -> None:
+        super().__init__(
+            name=name, domains=domains, operation=operation, children=children
+        )
+
+        self.original_operator: Operator
+        """Reference to the operator representing this operator at the current time amd
+        iterate.
+
+        This attribute is only available in operators representing previous time steps.
+
+        """
+
+        self._time_step_index: int = -1
+        """Time step index, starting with -1 (current time) and increasing for previous
+        time steps."""
+
+    @property
+    def prev_time(self) -> bool:
+        """True, if the operator represents a previous time-step."""
+        return True if self._time_step_index > -1 else False
+
+    @property
+    def time_step_index(self) -> Optional[int]:
+        """Returns the time step index this instance represents.
+
+        - None indicates this is an operator at the current time step
+        - 0 represents the first previous time step
+        - 1 represents the next time step further back in time
+        - ...
+
+        Note:
+            The public time step index is ``None``, if the operator represents the
+            current time.
+
+        """
+        return self._time_step_index if self._time_step_index >= 0 else None
+
+    def previous_timestep(
+        self: _TimeDependentOperator, steps: int = 1
+    ) -> _TimeDependentOperator:
+        """Returns a copy of the time-dependent operator with an advanced time-step
+        index.
+
+        Time-dependent operators do not invoke the recursion (like the base class),
+        but represent a leaf in the recursion tree.
+
+        Note:
+            You cannot create operators at the previous time step from operators which
+            are at some previous iterate. Use the :attr:`original_operator` instead.
+
+        Parameters:
+            steps: ``default=1``
+
+                Number of steps backwards in time.
```

Comment:
I agree it's confusing, but this is how it was implemented so far.
Maybe Operators at current time should have a default time step index 0 (which is never used, because the iterate index is used), and previous times should start with 1

Reply:
That could be an option, yes. Opinions, @keileg?

Reply:
See also discussion in https://github.com/pmgbergen/porepy/pull/1166#discussion_r1590274116

Reply:
Alright. Should I implement a raise error to prohibit users storing data at `time_step_index=0`?
I don't think there is a reason to ever use that key for storage, given the usage of `iterate_index` for the current time.

Reply:
Having thought more and discussed with @IvarStefansson, I suggest the following principles (written in terms of time_step_index, but it should be equally valid, and implemented, for iterate_index or whatever that is called):

1. We define the attribute time_step_index to tell where in the pp.STATE arrays to look for the values of this variable. This is one of two possible, reasonably consistent schemes (the other one is that the solution under construction is assigned index 0 for both time and index, but since that is an unknown state, this is somehow less useful).
2. The time step (iteration) being computed (call it the current one, or the one under construction) does not yet have a solution that can be looked up. To mark this, I actually prefer setting the index to None, since the alternative -1 actually can be used for array indexing, and that is not a nice bug to hunt down.
3. It should be possible to do some safeguarding in the get/set methods to make sure the above principles are followed.

Does this make sense? 

If so, follow-up question/comment to @vlipovac: While it would be great to get this super-clean, we also need to get this PR done, so that you can continue with your more important work on the compositional branch. It is therefore fine with me to be somewhat pragmatic in how polished these concepts are in the context of this PR, although the code should be working and not be more confusing. So, feel free to apply some cost-benefit thinking in this direction, and feel free to ask if you need guidance. Again, does that make sense?

Reply:
@keileg  I think I can spend 1-2 hours in getting the time_step_index right, but I was not able to understand your conclusion from various comments.

In your last comment, you say unification is the goal, here you say you prefer `time_step_index` to stay None for the solution under construction?